pkg/client/fetch.go

```diff
@@ -1 +1,265 @@
 package client
+
+import (
+	"context"
+	"errors"
+	"io"
+	"sync"
+	"time"
+
+	"github.com/cloudquery/cloudquery/internal/logging"
+	"github.com/cloudquery/cloudquery/pkg/client/state"
+	sdkdb "github.com/cloudquery/cq-provider-sdk/database"
+
+	"github.com/rs/zerolog"
+	gcodes "google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	"github.com/cloudquery/cq-provider-sdk/provider/diag"
+	"github.com/rs/zerolog/log"
+
+	"github.com/cloudquery/cloudquery/pkg/config"
+
+	"github.com/cloudquery/cloudquery/pkg/plugin"
+
+	"github.com/cloudquery/cloudquery/pkg/client/database"
+	"github.com/cloudquery/cloudquery/pkg/client/history"
+	"github.com/cloudquery/cloudquery/pkg/plugin/registry"
+	"github.com/cloudquery/cq-provider-sdk/cqproto"
+	"github.com/cloudquery/cq-provider-sdk/database/dsn"
+	"github.com/google/uuid"
+	"go.opentelemetry.io/otel/trace"
+)
+
+type ProviderInfo struct {
+	Provider registry.Provider
+	Config   *config.Provider
+}
+
+// FetchOptions is provided to the Client to execute a fetch on one or more providers
+type FetchOptions struct {
+	// UpdateCallback allows gets called when the client receives updates on fetch.
+	UpdateCallback FetchUpdateCallback
+	// Providers list of providers to call for fetching
+	ProvidersInfo []ProviderInfo
+	// Optional: Adds extra fields to the provider, this is used for history mode and testing purposes.
+	ExtraFields map[string]interface{}
+
+	History *history.Config
+}
+
+// FetchResponse is returned after a successful fetch execution, it holds a fetch summary for each provider that was executed.
+type FetchResponse struct {
+	ProviderFetchSummary map[string]ProviderFetchSummary
+}
+type fetchResult struct {
+	summary ProviderFetchSummary
+	diags   diag.Diagnostics
+}
+
+func parseDSN(storage database.Storage, cfg *history.Config) (string, error) {
+	if cfg == nil {
+		parsed, err := dsn.ParseConnectionString(storage.DSN())
+		if err != nil {
+			return "", err
+		}
+		return parsed.String(), nil
+	}
+	return history.TransformDSN(storage.DSN())
+}
+
+func Fetch(ctx context.Context, storage database.Storage, pm *plugin.Manager, opts FetchOptions) (res *FetchResponse, diagnostics diag.Diagnostics) {
+	fetchId, err := uuid.NewUUID()
+	if err != nil {
+		return nil, diag.FromError(err, diag.INTERNAL)
+	}
+	// set metadata we want to pass to
+	metadata := map[string]interface{}{"cq_fetch_id": fetchId.String()}
+	if opts.History != nil {
+		fd := opts.History.FetchDate()
+		log.Info().Str("fetch_date", fd.Format(time.RFC3339)).Stringer("fetch_id", fetchId).Msg("history enabled adding fetch date")
+		metadata["cq_fetch_data"] = fd
```

Comment:
I would keep the comment about the extrafield cq_fetch_date from the original (also typo: "data" vs "date")

Reply:
fixed